src/main/cpp/checkpoint/CheckPoint.cpp

```diff
@@ -686,12 +727,221 @@ void CheckPoint::WriteExpatriates(multiregion::ExpatriateJournal& journal, boost
 
 	hid_t dataset =
 	    H5Dcreate2(group, dsetname.c_str(), H5T_NATIVE_UINT, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
-	H5Dwrite(dataset, H5T_NATIVE_UINT, H5S_ALL, H5S_ALL, H5P_DEFAULT, &data.front());
+	H5Dwrite(dataset, H5T_NATIVE_UINT, H5S_ALL, H5S_ALL, H5P_DEFAULT, data.data());
 
 	H5Sclose(dataspace);
 	H5Dclose(dataset);
 	H5Gclose(group);
 }
 
+void CheckPoint::WriteVisitors(multiregion::VisitorJournal& journal, boost::gregorian::date date, std::size_t day)
+{
+	std::string datestr = to_iso_string(date);
+	htri_t exist = H5Lexists(m_file, datestr.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		hid_t temp = H5Gcreate2(m_file, datestr.c_str(), H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+		H5Gclose(temp);
+	}
+	hid_t group = H5Gopen2(m_file, datestr.c_str(), H5P_DEFAULT);
+
+	std::string dsetname = "Visitors";
+
+	std::vector<h_visitorType> data(journal.GetVisitorCount());
+
+	for (auto& days : journal.GetVisitors()) {
+		for (auto& place : days.second) {
+			for (auto p : place.second) {
+				h_visitorType i;
+				i.DaysLeft = days.first - day;
+				i.RegionID = place.first;
+				i.PersonIDHome = p.home_id;
+				i.PersonIDVisitor = p.visitor_id;
+				data.push_back(i);
+			}
+		}
+	}
+
+	hid_t newType = H5Tcreate(H5T_COMPOUND, sizeof(h_visitorType));
+
+	H5Tinsert(newType, "DaysLeft", HOFFSET(h_visitorType, DaysLeft), H5T_NATIVE_UINT);
+	H5Tinsert(newType, "RegionID", HOFFSET(h_visitorType, RegionID), H5T_NATIVE_UINT);
+	H5Tinsert(newType, "PersonIDHome", HOFFSET(h_visitorType, PersonIDHome), H5T_NATIVE_UINT);
+	H5Tinsert(newType, "PersonIDVisitor", HOFFSET(h_visitorType, PersonIDVisitor), H5T_NATIVE_UINT);
+
+	hsize_t dims = data.size();
+	hid_t dataspace = H5Screate_simple(1, &dims, nullptr);
+
+	hid_t dataset = H5Dcreate2(group, dsetname.c_str(), newType, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+	H5Dwrite(dataset, newType, H5S_ALL, H5S_ALL, H5P_DEFAULT, data.data());
+
+	H5Tclose(newType);
+	H5Sclose(dataspace);
+	H5Dclose(dataset);
+	H5Gclose(group);
+}
+
+multiregion::ExpatriateJournal CheckPoint::LoadExpatriates(const Population& pop, boost::gregorian::date date)
+{
+	multiregion::ExpatriateJournal result;
+	std::string dsetName = to_iso_string(date) + "/Expatriates";
+	htri_t exist = H5Lexists(m_file, dsetName.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		return result;
+	}
+	hid_t dset = H5Dopen2(m_file, dsetName.c_str(), H5P_DEFAULT);
+
+	hid_t dspace = H5Dget_space(dset);
+
+	hsize_t dims;
+	H5Sget_simple_extent_dims(dspace, &dims, nullptr);
+
+	std::vector<unsigned int> data(dims);
+
+	H5Dread(dset, H5T_NATIVE_UINT, H5S_ALL, H5S_ALL, H5P_DEFAULT, data.data());
+
+	H5Sclose(dspace);
+	H5Dclose(dset);
+
+	for (auto& i : data) {
```

Comment:
I don't think that what this loop does is okay. I have two main concerns:

  1. It looks for expatriates (i.e., people who are _not_ present in the population) in the population. So `p.GetId() == i` should *never* evaluate to `true`. In effect this algorithm does nothing.

  2. It does so very inefficiently: this is an `O(n^2)` algorithm and `n` is usually huge.